Fill.py

```diff
@@ -921,15 +921,32 @@ def failed(warning: str, force: typing.Union[bool, str]) -> None:
             multiworld.random.shuffle(items)
             count = 0
             err: typing.List[str] = []
-            successful_pairs: typing.List[typing.Tuple[Item, Location]] = []
+            successful_pairs: typing.List[typing.Tuple[int, Item, Location]] = []
+            claimed_indices: typing.Set[typing.Optional[int]] = set()
             for item_name in items:
-                item = multiworld.worlds[player].create_item(item_name)
+                index_to_delete: typing.Optional[int] = None
+                if from_pool:
+                    try:
+                        # If from_pool, try to find an existing item with this name & player in the itempool and use it
+                        index_to_delete, item = next(
+                            (i, item) for i, item in enumerate(multiworld.itempool)
+                            if item.name == item_name and item.player == player and i not in claimed_indices
```

Comment:
This condition could be rearranged to be quicker and likely short circuit sooner
```suggestion
                            if item.player == player and item.name == item_name and i not in claimed_indices
```